### ISSUE TITLE
Mano.ini

### ISSUE THREAD

**Issue by Gavin-Holt** [mougino/MaNo#2]
Hi,

Would you consider using a local Mano.ini file to allow user customization? I prefer local ini files as I run all my software from a portable USB. This would allow users to change editor settings and full control over generated HTML, especially important for those unwilling to run remote java scripts. Below are some ideas that might be included:

```
[html]
<!DOCTYPE html>
<html lang="en">

[Head]
<meta name="generator" content="MaNo the Markdown Notepad">
<meta name="identifier-url" content="http://mougino.free.fr">
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<link href="http://mougino.free.fr/markdown1.css" media="all" rel="stylesheet">
<link href="http://mougino.free.fr/markdown2.css" media="all" rel="stylesheet">
<meta name="viewport" content="initial-scale=1.0,user-scalable=no,maximum-scale=1,width=device-width" media="(device-height: 568px)">
<title>$FileName.md</title>
<link rel="icon" type="image/x-icon" href="http://mougino.free.fr/MaNo.ico">
<meta name="theme-color" content="#1e2327">

[scripts]
<script src="http://mougino.free.fr/markdown1.js"></script>
<script src="http://mougino.free.fr/markdown2.js"></script>

[Editor]
FontName="Courier New"
FontSize=11
FontForeColor=
FontBackColor=
ShowLineNumbers=1
LineNumbersWidth=44
LineNumberFontName="Courier New"
LineNumberFontSize=11
LineNumberFontForeColor=95,211,188
LineNumberFontBackColor=
Charset=0
StatusBar=1
HTMLProtect=1
```
I couldn't find all the default values in the source. As you can see I have sneaked in some suggestions for added functionality :)

- Line numbers - show/hide , margin is sizable, and font/fore/back colors can be set.
- StatusBar - show/hide
- [HTMLProtect]( https://github.com/mougino/MaNo/blob/df1b5b7159a7989f940d6b079127f2d0f42fbdbe/MaNo.bas#L2429) - allow users to enter tags 

P.S. What are markdown1.js and markdown2.js doing?

Kind Regards Gavin Holt

**Comment by mougino** [comment]
Hi Gavin, there are a lot of ideas here, let me try to answer one by one:

1. local ini file instead of registry

Great idea but how does the program decide when it uses registry, and when it uses ini?
I would propose to give precedence to an existing local .ini file, with a fallback on registry.
But then: when/how to dump an ini file for the user to modify?

2. ini sections

[html], [head] and [scripts] sections should not be in a .ini, as they contain html code.
If you open mano.exe as a zip (right-clic > 7zip > open archive) then in .rsrc\RCDATA you have 4 files numbered 6 to 9. These are the html resources, as such:
```
    6 RCDATA DISCARDABLE "markdown.head"
    7 RCDATA DISCARDABLE "markdown.foot"
    8 RCDATA DISCARDABLE "markdown.svg"
    9 RCDATA DISCARDABLE "emojis.lst"
```
I made an archive of these files [available here](http://mougino.free.fr/tmp/manores.zip), feel free to dig in.
I propose to have a mano developer menu (e.g. via the F12 key) that will allow to dump these files locally. Then the user can modify them, and I propose to modify the program to take local files as precedence over embedded resources.

3. editor ini keys

I don't see any big issue in principle to allow dynamic font family, size, foreground and background colors. Change should be relatively easy so I'll try next week (I'm afk this weekend).

I don't feel comfortable allowing to modify the **status bar** and **lineNb column** on the other hand...
The **status bar** shows important information such as unsaved state, Insert key state, etc. The user needs those.
As for the **lineNb column**, if you use Ctrl+wheel up/down for zooming, you'll see it is resized on the fly. That was a huge pain to implement correctly. If I try to handle a case where the column does not exist, I open the door to possibly dangerous side effects. And showing/hiding is one thing, but dynamic font, size etc: this cannot be done without harm, I'd have to use a certain hack to calculate the size of labels dinamically, and size the column accordingly. I have done that for another program, it was a huge pain, we'd consider hundreds of lines of code and possible memory leak source to achieve the result. I don't want my lean editor to become a bomber engine...

4. HTMLProtect

Allowing users to write html would be a powerful feature, but basically break compatibility with GFM (GitHub Flavored Markdown).
mano is a markdown editor, not an html editor. It *should* support certain html tags as defined here: [HTML Tags You Can Use on GitHub](https://gist.github.com/seanh/13a93686bf4c2cb16e658b3cf96807f2) but I've never tested so I'm not sure.
Also, if we did allow html, an "all or nothing" switch in an ini file is not the way to go, we'd have to think of a better way.

Let me try to think more on this fourth feature. It's an interesting one for sure.

5. markdown1.js and markdown2.js

These are aggregated and minified files of the different javascript sources used by Github for their HTML Markdown rendering. These file were made manually via retro-engineering, I don't have the detail anymore of how I proceeded, sorry (it was almost 5 years ago).

6. remote java scripts

You mention that a reason for modifying the html header and footer is for those unwilling to run remote java scripts.
Couldn't we add an option to dump local .js and .css and use those in the generated htmls instead?

**Comment by mougino** [comment]
Making progress ... ;)
![mano_ini_custo](https://user-images.githubusercontent.com/2697547/141781246-6d1dacde-aca6-497a-bf99-09a0dbf7aaba.jpg)

**Comment by mougino** [comment]
1. local ini file instead of registry

100% implemented. You can dump an original MaNo.ini to modify and play with through the new **Developer Mode** (`F12`):

![mano_dev_mode](https://user-images.githubusercontent.com/2697547/141834027-cd6edc6a-fcbd-4db1-9912-d526cf3a386c.jpg)

2. ini sections

HTML header (css, favicon) and footer (scripts) are handled via their respective files dumped through the **Developer Mode** (`F12`). When generating html, these files are taken in priority, so if you modify them the resulting html will be modified consequently.

![mano_html](https://user-images.githubusercontent.com/2697547/141837826-27214509-d4e9-4d05-879c-0311e461a242.jpg)

3. editor ini keys

100% implemented. You can now change **editor** font, font size, foreground and background colors ; and **line number column** font, font size, foreground and background colors, as well as visibility (show or hide). The **status bar** stays at all time.

4. HTMLProtect

Not yet worked on.

5. remote js and css

**Local js, css, and favicon** are dumped through the **Developer Mode** (`F12`). If you generate an html with those locally, they are now referenced locally instead of pointing to the remote ones from http;//mougino.free.fr

![mano_reference](https://user-images.githubusercontent.com/2697547/141838151-34d37c28-178e-4f95-a9da-14ff42d36e5f.jpg)

**Conclusion**

I need to do some testing before committing my changes and release **1.3.0**.
That's for the serious part.

But the new font/color possibility make me want to make a **MaNo skin manager** and have some fun 😉 

Tell me if you're interested in an intermediate release to play with and I'll send you a link.

NB: MaNo.exe is now 1 MB vs. 275 KB for 1.2.0.

**Comment by mougino** [comment]
Just realizing I should probably rename "Developer Mode" to "Portable Mode"...

**Comment by mougino** [comment]
A preview version of MaNo 1.3.0 + Skin Manager [is available here](http://mougino.free.fr/tmp/MaNo1.3.0+SkinMgr.zip)

**Comment by mougino** [comment]
Fixed with commit [#ac2d457](https://github.com/mougino/MaNo/commit/ac2d457a13745e28a38b5b1e3a8a6efb0f0c4276)
